src/foam/nanos/ruler/predicate/MQL.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.ruler.predicate',
+  name: 'MQL',
+  extends: 'foam.mlang.predicate.AbstractPredicate',
+  implements: ['foam.core.Serializable'],
+
+  documentation: `A predicate that converts mql query to predicate and evaluates object`,
+
+  javaImports: [
+    'foam.core.FObject',
+    'static foam.mlang.MLang.*'
+  ],
+  properties: [
+    {
+      class: 'Boolean',
+      name: 'isNew',
+      value: true
+    },
+    {
+      class: "String",
+      name: 'query'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'f',
+      javaCode: `
+        if ( getIsNew() ) {
+          FObject nu  = (FObject) NEW_OBJ.f(obj);
+          return MQL(getQuery()).f(nu);
```

Comment:
This is going to create/compile/specialize a new MQL for each call to f(), which is going to be very slow.

Reply:
Let's remove the MQL constructor and add a caching create() method, like many of the other parsers do. On the JS side we can make it a Multiton on the 'query' property. See docs in Multiton.js.